In [5]:
import pyspark
from pyspark.sql import SparkSession
import pandas as pd

### Question 3. Connecting to the Kafka server

In [1]:
import json
import time 

from kafka import KafkaProducer

def json_serializer(data):
    return json.dumps(data).encode('utf-8')

server = 'localhost:9092'

producer = KafkaProducer(
    bootstrap_servers=[server],
    value_serializer=json_serializer
)

In [2]:
# check output of this
producer.bootstrap_connected()

True

Answer is True

### Question 4. Sending data to the stream


In [3]:
t0 = time.time()

topic_name = 'test-topic'

send_msg_start = time.time()
for i in range(10):
    message = {'number': i}
    producer.send(topic_name, value=message)
    print(f"Sent: {message}")
    time.sleep(0.05)
send_msg_end = time.time()

flush_start = time.time()
producer.flush()
flush_end = time.time()

t1 = time.time()
print(f'all took {(t1 - t0):.2f} seconds')
print(f'sending messages took {(send_msg_end - send_msg_start):.2f} seconds')
print(f'flushing took {(flush_end - flush_start):.2f} seconds')

Sent: {'number': 0}
Sent: {'number': 1}
Sent: {'number': 2}
Sent: {'number': 3}
Sent: {'number': 4}
Sent: {'number': 5}
Sent: {'number': 6}
Sent: {'number': 7}
Sent: {'number': 8}
Sent: {'number': 9}
all took 0.50 seconds
sending messages took 0.50 seconds
flushing took 0.00 seconds


Answer:
the entire cell took 0.5 seconds. it spent most of the time in sending the message.

### Sending the taxi data


In [6]:
green_df = pd.read_csv('green_tripdata_2019-10.csv.gz', low_memory=False)

In [7]:
# reduce columns
green_df = green_df[[
    'lpep_pickup_datetime',
    'lpep_dropoff_datetime',
    'PULocationID',
    'DOLocationID',
    'passenger_count',
    'trip_distance',
    'tip_amount'
]]

In [8]:
green_df.head()

lpep_pickup_datetime lpep_dropoff_datetime  PULocationID  DOLocationID  \
0  2019-10-01 00:26:02   2019-10-01 00:39:58           112           196   
1  2019-10-01 00:18:11   2019-10-01 00:22:38            43           263   
2  2019-10-01 00:09:31   2019-10-01 00:24:47           255           228   
3  2019-10-01 00:37:40   2019-10-01 00:41:49           181           181   
4  2019-10-01 00:08:13   2019-10-01 00:17:56            97           188   

   passenger_count  trip_distance  tip_amount  
0              1.0           5.88        0.00  
1              1.0           0.80        0.00  
2              2.0           7.50        0.00  
3              1.0           0.90        0.00  
4              1.0           2.52        2.26

In [9]:
green_df.shape

(476386, 7)

### Question 5: Sending the Trip Data


In [10]:
TOPIC_NAME = 'green-trips'

start = time.time()
send_msg_start = time.time()
for row in green_df.itertuples(index=False):
    row_dict = {col: getattr(row, col) for col in row._fields}
    # print(row_dict) # REMOVING PRINT STATEMETN BECCAUSE IT IS CAUSING 'IOPub data rate exceeded'

    message = row_dict
    producer.send(TOPIC_NAME, value=message)

send_msg_end = time.time()

flush_start = time.time()
producer.flush()
flush_end = time.time()

end = time.time()

print(f'all took {(end - start):.2f} seconds')
print(f'sending messages took {(send_msg_end - send_msg_start):.2f} seconds')
print(f'flushing took {(flush_end - flush_start):.2f} seconds')

all took 34.48 seconds
sending messages took 30.63 seconds
flushing took 3.85 seconds


How much time in seconds did it take? (You can round it to a whole number)
36 Seconds


### Creating the PySpark consumer

In [11]:
import pyspark
from pyspark.sql import SparkSession

pyspark_version = pyspark.__version__
kafka_jar_package = f"org.apache.spark:spark-sql-kafka-0-10_2.12:{pyspark_version}"

spark = SparkSession \
    .builder \
    .master("local[*]") \
    .appName("GreenTripsConsumer") \
    .config("spark.jars.packages", kafka_jar_package) \
    .getOrCreate()

24/03/15 21:11:22 WARN Utils: Your hostname, coy-pc resolves to a loopback address: 127.0.1.1; using 172.19.138.153 instead (on interface eth0)
24/03/15 21:11:22 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
:: loading settings :: url = jar:file:/home/frean/spark/spark-3.3.2-bin-hadoop3/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/frean/.ivy2/cache
The jars for the packages stored in: /home/frean/.ivy2/jars
org.apache.spark#spark-sql-kafka-0-10_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-b4abb0ec-8cba-4572-aa9a-66aab3946726;1.0
	confs: [default]
	found org.apache.spark#spark-sql-kafka-0-10_2.12;3.3.2 in central
	found org.apache.spark#spark-token-provider-kafka-0-10_2.12;3.3.2 in central
	found org.apache.kafka#kafka-clients;2.8.1 in central
	found org.lz4#lz4-java;1.8.0 in central
	found org.xerial.snappy#snappy-java;1.1.8.4 in central
	found org.slf4j#slf4j-api;1.7.32 in central
	found org.apache.hadoop#hadoop-client-runtime;3.3.2 in central
	found org.spark-project.spark#unused;1.0.0 in central
	found org.apache.hadoop#hadoop-client-api;3.3.2 in central
	found commons-logging#commons-logging;1.1.3 in central
	found com.google.code.findbugs#jsr305;3.0.0 in central
	found org.apache.commons#commons-pool2;2.11.1 in central
:: resolu

24/03/15 21:11:24 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


In [12]:
green_stream = spark \
    .readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", "localhost:9092") \
    .option("subscribe", "green-trips") \
    .option("startingOffsets", "earliest") \
    .load()

In [37]:
def peek(mini_batch, batch_id):
    first_row = mini_batch.take(1)

    if first_row:
        print(first_row[0])

query = green_stream.writeStream.foreachBatch(peek).start()

24/03/15 21:00:14 WARN ResolveWriteToStream: Temporary checkpoint location created which is deleted normally when the query didn't fail: /tmp/temporary-bacf050b-fb1f-47cc-be9d-872fa65cbdd7. If it's required to delete it under any circumstances, please set spark.sql.streaming.forceDeleteTempCheckpointLocation to true. Important to know deleting temp checkpoint folder is best effort.
24/03/15 21:00:14 WARN ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
Row(key=None, value=bytearray(b'{"lpep_pickup_datetime": "2019-10-01 00:26:02", "lpep_dropoff_datetime": "2019-10-01 00:39:58", "PULocationID": 112, "DOLocationID": 196, "passenger_count": 1.0, "trip_distance": 5.88, "tip_amount": 0.0}'), topic='green-trips', partition=0, offset=0, timestamp=datetime.datetime(2024, 3, 15, 20, 55, 39, 930000), timestampType=0)
Row(lpep_pickup_datetime='2019-10-01 00:26:02', lpep_dropoff_datetime='2019-10-01 00:39:58', PULocationID=112

In [38]:
query.stop()

### Question 6. Parsing the data


In [13]:
from pyspark.sql import types

schema = types.StructType() \
    .add("lpep_pickup_datetime", types.StringType()) \
    .add("lpep_dropoff_datetime", types.StringType()) \
    .add("PULocationID", types.IntegerType()) \
    .add("DOLocationID", types.IntegerType()) \
    .add("passenger_count", types.DoubleType()) \
    .add("trip_distance", types.DoubleType()) \
    .add("tip_amount", types.DoubleType())

In [14]:
from pyspark.sql import functions as F

green_stream = green_stream \
  .select(F.from_json(F.col("value").cast('STRING'), schema).alias("data")) \
  .select("data.*")

In [15]:
def peek(mini_batch, batch_id):
    first_row = mini_batch.take(1)

    if first_row:
        print(first_row[0])

query = green_stream.writeStream.foreachBatch(peek).start()

24/03/15 21:12:16 WARN ResolveWriteToStream: Temporary checkpoint location created which is deleted normally when the query didn't fail: /tmp/temporary-713de3e2-fc0f-4119-bf08-84119d3acf55. If it's required to delete it under any circumstances, please set spark.sql.streaming.forceDeleteTempCheckpointLocation to true. Important to know deleting temp checkpoint folder is best effort.
24/03/15 21:12:16 WARN ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
Row(lpep_pickup_datetime='2019-10-01 00:26:02', lpep_dropoff_datetime='2019-10-01 00:39:58', PULocationID=112, DOLocationID=196, passenger_count=1.0, trip_distance=5.88, tip_amount=0.0)


In [16]:
query.stop()

How does the record look after parsing? Copy the output
Row(lpep_pickup_datetime='2019-10-01 00:26:02', lpep_dropoff_datetime='2019-10-01 00:39:58', PULocationID=112, DOLocationID=196, passenger_count=1.0, trip_distance=5.88, tip_amount=0.0)

without schema:
Row(key=None, value=bytearray(b'{"lpep_pickup_datetime": "2019-10-01 00:26:02", "lpep_dropoff_datetime": "2019-10-01 00:39:58", "PULocationID": 112, "DOLocationID": 196, "passenger_count": 1.0, "trip_distance": 5.88, "tip_amount": 0.0}'), topic='green-trips', partition=0, offset=0, timestamp=datetime.datetime(2024, 3, 15, 20, 55, 39, 930000), timestampType=0)


### Question 7: Most popular destination


In [ ]:
# did not do the workshop 2 yet, i don't know how to proceed :(